File: app/src/main/java/ezlife/movil/oneparkingapp/activities/LoginActivity.java
package ezlife.movil.oneparkingapp.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import ezlife.movil.oneparkingapp.R;
import ezlife.movil.oneparkingapp.databinding.LoginBinding;
import ezlife.movil.oneparkingapp.net.api.login.ClientRes;
import ezlife.movil.oneparkingapp.net.api.login.LoginApi;
import ezlife.movil.oneparkingapp.net.http.HttpError;
import ezlife.movil.oneparkingapp.util.CryptoJs;
import ezlife.movil.oneparkingapp.util.Preference;
import ezlife.movil.oneparkingapp.util.Qk;
import ezlife.movil.oneparkingapp.util.Settings;

public class LoginActivity extends AppCompatActivity implements CryptoJs.OnEncryptAES, LoginApi.OnLoginListener {

    private static final int REQUEST_REG = 101;

    LoginBinding binding;
    LoginApi api;

    SharedPreferences preferences;
    String city;

    CryptoJs cryptoJs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login);
        binding.setHandler(this);

        preferences = getSharedPreferences(Preference.NAME, MODE_PRIVATE);
        city = preferences.getString(Preference.CITY, null);
        binding.setCity(city);

        cryptoJs = new CryptoJs(this);

        api = new LoginApi(this, Qk.makeLoading(this));
    }

    public void clear() {
        city = null;
        binding.setCity(null);
    }

    public void goToRegister() {
        Intent intent = new Intent(this, RegisterActivity.class);
        startActivityForResult(intent, REQUEST_REG);
    }

    public void login() {
        String usr = Qk.getText(binding.usr);
        String pass = Qk.getText(binding.pass);
        if (usr.equals("") || pass.equals("")) {
            Qk.showToast(this, R.string.login_form);
        } else {
            if (city == null) {
                city = binding.city.getSelectedItem().toString();
            }

            String auth = usr + "_&&_" + pass + "_&&_" + System.currentTimeMillis();
            cryptoJs.encryptAES(auth, Settings.SECRET, this);
        }
    }

    @Override
    public void onEncryptedAES(String aes) {
        api.login(aes, this);
    }

    @Override
    public void onLogin(boolean success, ClientRes.Client usuario, int error) {
        if (success) {
            Intent intent;
            SharedPreferences.Editor editor = preferences.edit();
            editor.putString(Preference.CITY, city);
            editor.apply();
            if (usuario.isValidado()) {
                if(usuario.getVehiculos().size() > 0)
                    intent = new Intent(this, MapActivity.class);
                else{
                    intent = new Intent(this, AddCarActivity.class);
                    intent.putExtra(AddCarActivity.EXTRA_FIRST_TIME, true);
                }
            } else {
                intent = new Intent(this, PassActivity.class);
            }
            startActivity(intent);
            finish();

        } else if (!success && error == HttpError.NO_ERROR) {
            Qk.showToast(this, R.string.login_fail);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if(requestCode == REQUEST_REG){
            if(resultCode == RESULT_OK){
                Intent intent = new Intent(this, AddCarActivity.class);
                intent.putExtra(AddCarActivity.EXTRA_FIRST_TIME, true);
                startActivity(intent);
                finish();
            }
        }
    }
}

File: app/src/main/java/ezlife/movil/oneparkingapp/activities/LoginActivity.kt
package ezlife.movil.oneparkingapp.activities

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.AssetManager
import android.databinding.BindingAdapter
import android.databinding.DataBindingUtil
import android.graphics.Typeface
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.widget.TextView
import ezlife.movil.oneparkingapp.R
import ezlife.movil.oneparkingapp.databinding.LoginBinding
import ezlife.movil.oneparkingapp.db.Car
import ezlife.movil.oneparkingapp.db.CarDao
import ezlife.movil.oneparkingapp.db.DB
import ezlife.movil.oneparkingapp.providers.UserProvider
import ezlife.movil.oneparkingapp.util.Preference
import ezlife.movil.oneparkingapp.util.asyncUI
import ezlife.movil.oneparkingapp.util.await
import ezlife.movil.oneparkingapp.util.text

class LoginActivity : AppCompatActivity() {

    lateinit var binding: LoginBinding
    val provider: UserProvider by lazy { UserProvider(this, makeLoading()) }
    val dao:CarDao by lazy{ DB.con.carDao() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login)
        binding.handler = this
        checkPermission()
    }

    fun login() {
        val usr = "${binding.usr.text()}"
        val pass = "${binding.pass.text()}"
        if (usr == "" || pass == "") {
            toast(R.string.login_form)
            return
        }
        provider.login(usr, pass) { user ->

            if(user.validado){
                if(user.vehiculos.isNotEmpty()) {
                    goToMap(user.vehiculos)
                }else
                    startActivity<AddCarActivity>(AddCarActivity.EXTRA_FIRST_TIME to true)

            }else
                startActivity<PassActivity>(PassActivity.EXTRA_CARS to user.vehiculos.isNotEmpty())
            finish()
        }
    }

    fun goToMap(cars:List<Car>) = asyncUI{
        cars[0].selected = true
        await { dao.insertList(cars) }
        savePreference(Preference.USER_LOGGED to true)
        startActivity<MapActivity>()
    }

    fun clear() {
        binding.city = null
    }

    fun goToRegister(){
        startActivityForResult<RegisterActivity>(101)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if(requestCode == 101){
            if(resultCode == Activity.RESULT_OK){
                startActivity<AddCarActivity>(AddCarActivity.EXTRA_FIRST_TIME to true)
                finish()
            }
        }
    }

    fun checkPermission() {

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
                || ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED
                || ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                ) {

            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE
                    , Manifest.permission.CAMERA
                    , Manifest.permission.ACCESS_FINE_LOCATION), 101)

        }
    }

    companion object{
        @JvmStatic
        @BindingAdapter("app:roboto")
        fun loadRoboto(view: TextView, name:String){
            val assetManager: AssetManager = view.context.assets
            val typeface: Typeface = Typeface.createFromAsset(assetManager,"fonts/$name.ttf")
            view.typeface = typeface
        }
    }
}